Fake: copy over grad attribute

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82593
* #82571
* #82337
* #82449

Copy over the grad attribute... let me any suggestions of other tests/opinfos/crossrefs I should be doing to make this more comprehensive.

eellison committed:
Fake: copy over grad attribute

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -137,6 +137,8 @@ def from_real_tensor(self, fake_mode, t):
         if type(t) is torch.nn.Parameter:
             out = torch.nn.Parameter(out, requires_grad=out.requires_grad)  # type: ignore[assignment]
+        if safe_is_leaf(t) and t.grad is not None:
+            out.grad = self.from_real_tensor(fake_mode, t.grad)
```
what if we have grad on a non-leaf tensor 🤔

eellison replied:
TBH, i'm not really sure...

ezyang replied:
I think I'd just drop the leaf check

```
>>> torch.zeros(2, requires_grad=True)
tensor([0., 0.], requires_grad=True)
>>> x = torch.zeros(2, requires_grad=True)
>>> y = x * 2
>>> y.grad = torch.zeros(2)
>>> y
tensor([0., 0.], grad_fn=<MulBackward0>)
>>> y.grad
tensor([0., 0.])
```

eellison committed:
Update on "Fake: copy over grad attribute"


Copy over the grad attribute... let me any suggestions of other tests/opinfos/crossrefs I should be doing to make this more comprehensive. 

[ghstack-poisoned]

eellison committed:
Update on "Fake: copy over grad attribute"


Copy over the grad attribute... let me any suggestions of other tests/opinfos/crossrefs I should be doing to make this more comprehensive. 

[ghstack-poisoned]

eellison committed:
Update on "Fake: copy over grad attribute"


Copy over the grad attribute... let me any suggestions of other tests/opinfos/crossrefs I should be doing to make this more comprehensive. 

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
